Add a Python version of the Fortran code

DWesl opened the pull request:
@karlwx, @ShawnMurd: here's a draft python version of the Fortran code.  So far it's almost a line-for-line translation of the Fortran code, so the circulation calculation will likely be slow.  I haven't run it yet to see if it's slow enough to cause problems.

DWesl committed:
ENH: Add a python version of the Fortran program.

At this point it's just a line-by-line translation.
The circulation calculatin will be very slow.
I can speed that up some by Cythonizing circ.py and interp.py,
but that seems out-of-scope.
I could also speed that up some by vectorizing the interpolation
and circulation calculation.  SciPy has vectorized interpolation;
not sure if I want to add that dependency.

DWesl committed:
STY: Run isort on python files.

DWesl committed:
STY: Run black on python code.

Try to get code somewhat more readable.

DWesl committed:
STY: Fix warnings from flake8.

Not many left at this point, but some.
Mostly reminders to add examples.

DWesl committed:
STY: Run pydocstringformatter on files.

Had to lengthen module docstrings so I wouldn't have other tools complaining.

DWesl committed:
STY: Fix most mypy warnings.

There's still a few where NumPy doesn't type functions in np.char.
I would also like to specify array sizes, but I haven't figured out how to do that yet.

DWesl committed:
STY: Fix other formatter warnings.

DWesl committed:
BUG: Get python version of code working.

Haven't checked if it runs to completion.
Haven't checked if it does the right thing.
It took 50ish seconds before I stopped it last time.

DWesl committed:
BUG: Get fortran code compiling with few warnings.

DWesl commented:
The Fortran version takes less than a second, while the python version could take multiple minutes.  That needs to be cut down a bit (ideally to ten seconds or less) before being used in a workshop.

UPDATE: I timed the Fortran vesion as running in fifteen seconds including compilation, and the Python version as taking over an hour.

DWesl committed:
STY: Convert Fortran code to use modules.

I was hoping this would avoid the random failures on file open.
This does not appear to have helped.

DWesl committed:
STY: Convert getsizes to use the F90 interface.

Also adds error checking.
I wanted to see where in this function the program was aborting.
It seems to be somewhere in nf90_open.
Also the program still returns 0 exit status, which is strange.

ShawnMurd commented:
I'm not surprised that the circulation code takes so long to run in Python. I did create a Pythonic version of it a while ago, but I cheated and used Numba's JIT compiler to speed it up so that it wasn't that much slower than the Fortran version. That's one approach you could use to speed things up. All you would need to do is add the `@jit` decorator before the function definition.

I wonder if maybe it would be easier to keep the circulation computation in Fortran, but use Python to plot the results. That would keep both the Fortran and Python components, and would perhaps better represent a typical research workflow (i.e., use Fortran for the heavy computations, then use Python for data perusal and visualization). Just a thought. You're leading the workshop this year, so you do what you feel is best!

ShawnMurd commented:
As a follow-up, is the idea to have the Python/Fortran code be a black box? If that's the case, you can use whatever hacks you want to speed it up. When I did the workshop last year, I don't think we even opened any of the Fortran files. The main objective was just to show how to compile and run a Fortran program in Linux if someone shares a Fortran program with you, not how to actually code in Fortran.

DWesl committed:
ENH: Use scipy interpn instead of home-rolled.

Still an order of magnitude slower than Fortran,
but it now only takes a minute instead of over an hour
(or freezing the computer).

DWesl committed:
PERF: Use numexpr for the integral calculation.

This doesn't seem to have sped things any.

DWesl committed:
ENH: Convert back to home-rolled interpolation.

Vectorized now, and closer to the Fortran for it.
Twenty seconds faster than SciPy's interpolation;
I can probably speed that up a bit with numexpr.

DWesl committed:
PERF: Use numexpr for the interpolation.

This brings a factor-of-three improvement,
likely from eliminating copies and only doing one pass through inputs,
since NumPy's started adding SIMD instructions to its UFunc loops.

This is now basically as fast as it needs to be,
although still an order of magnitude slower than Fortran.

DWesl commented:
> I'm not surprised that the circulation code takes so long to run in Python. I did create a Pythonic version of it a while ago, but I cheated and used Numba's JIT compiler to speed it up so that it wasn't that much slower than the Fortran version. That's one approach you could use to speed things up. All you would need to do is add the `@jit` decorator before the function definition.

Numba is a bit hard for me to work with.  I tried doing multiple interpolations at once using `scipy.interpolate.interpn`, which took about a minute, then I tried doing that with the Fortran interpolation algorithm, which took about forty seconds, then I tried using numexpr with the Fortran implementation (not the python one I had before), which took the time down to fifteen seconds.

Numexpr is just as much of a cheat as Numba, but I've managed to get it to compile on my setup.  I could change `circ.py` and `interp.py` into Cython files while still keeping the Pythonic syntax, but I'd prefer to avoid the compile step.

> I wonder if maybe it would be easier to keep the circulation computation in Fortran, but use Python to plot the results. That would keep both the Fortran and Python components, and would perhaps better represent a typical research workflow (i.e., use Fortran for the heavy computations, then use Python for data perusal and visualization). Just a thought. You're leading the workshop this year, so you do what you feel is best!

It would be possible to wrap `circ.f` and `interp.f` into a Python "C" extension module with `f2py`, but I don't know that I could get that working by Monday.  It would show best practices for project code, as defined by R's tidyverse proponents; I might look into that this weekend.

Either way, there would need to be modifications to the presentation to walk people through 
```bash
module load python/anaconda3_...
conda create -n linux-intro numpy netCDF4 numexpr ...
source activate linux-intro
```
and at least a general idea of what each of those commands does.

> As a follow-up, is the idea to have the Python/Fortran code be a black box? If that's the case, you can use whatever hacks you want to speed it up. When I did the workshop last year, I don't think we even opened any of the Fortran files. The main objective was just to show how to compile and run a Fortran program in Linux if someone shares a Fortran program with you, not how to actually code in Fortran.

Yeah, that's probably what I'll do this year, but it took me embarrassingly long in the process of setting up to remember that (to read through the presentation where it says that).

ShawnMurd commented:
> 
> Numba is a bit hard for me to work with. I tried doing multiple interpolations at once using `scipy.interpolate.interpn`, which took about a minute, then I tried doing that with the Fortran interpolation algorithm, which took about forty seconds, then I tried using numexpr with the Fortran implementation (not the python one I had before), which took the time down to fifteen seconds.
> 

Wow, I'm impressed that you got it down that low! 

> 
> It would be possible to wrap `circ.f` and `interp.f` into a Python "C" extension module with `f2py`, but I don't know that I could get that working by Monday. It would show best practices for project code, as defined by R's tidyverse proponents; I might look into that this weekend.
> 

Yup, f2py can be a pain. Probably best to not show this during the workshop.

> 
> Yeah, that's probably what I'll do this year, but it took me embarrassingly long in the process of setting up to remember that (to read through the presentation where it says that).

Yeah, keep in mind that the main goal is just to get people comfortable with a Linux terminal. It's probably best not to do anything too fancy (or try to teach too much), otherwise, that might detract from the main goal.

DWesl committed:
ENH: Get f2py working with circ.f and interp.f.

I can't quite get it working with scikit-build,
but that seems to be the install step, not the build step.

DWesl committed:
ENH: Restructure main.py to use function.

Allows re-use in package call or console_scripts entry point.
This should be before the 'Add f2py' commit.

DWesl committed:
DEV: Add a CMake build interface.

I can't auto-detect netCDF, so it's not as versatile
as the autotools build.

DWesl committed:
ENH: Get Python code working as package.

F2PY wrappers of circ and interp take python execution time
down to a second.

DWesl committed:
STY: Update indentation and remove dead code.

DWesl committed:
Merge branch 'main' into python-version

DWesl committed:
DEV: Build interp.f90 befor circ.f90 so module is available.

DWesl committed:
DEV: Build script: ensure modules available before use.

DWesl committed:
DEV: Update build script to reflect name change.

interp.f and circ.f got renamed to be .f90 files.
This made it easier to get f2py working.

DWesl committed:
DEV: Clean up built modules.

The compilation process produces .mod files.
The cleanup process should remove them.

DWesl committed:
CI: Add a CMake CI run.

The test step might cause problems; we will see.

DWesl committed:
DEV: Only explicitly link to netcdff and netcdf in CMake

Apparently I don't need to link to hdf on Cygwin, and definitely don't
on Linux.

DWesl committed:
Merge branch 'python-version' of github.com:DWesl/linux_ws_exercise into python-version

DWesl committed:
CI: Fix CMake build directory for CTest.

I forgot to change it in the other steps, so I may as well change this
to match.

DWesl committed:
CI: Add a test of the Python implementation.

At this point, it's mostly just a check that it actually installs.

DWesl committed:
CI: Start python tests in correct directory.

It was starting in the parent directory.

Having projects be in a subdirectory rather than root is kind of
annoying.

DWesl committed:
DEV: Add NumPy to build requirements.

I use f2py, so I kind of need it.

DWesl committed:
DEV: Fix scikit-build compile order.

interp.f90 and circ.f90 need to be compiled in that order so the
necessary modules exist.

DWesl commented:
> > It would be possible to wrap `circ.f` and `interp.f` into a Python "C" extension module with `f2py`, but I don't know that I could get that working by Monday. 
> 
> Yup, f2py can be a pain. Probably best to not show this during the workshop.

I got the `f2py` wrapper working this morning, and that implementation keeps the 1 s time scale; I think the other code takes longer mostly because it makes more passes through the data (the Fortran version goes through u and v once each; the numexpr version does over ten).  

I have a draft Cython version that should be similar to the Fortran version, but I can't convince the build system that it's an extension module.
> 
> > Yeah, that's probably what I'll do this year, but it took me embarrassingly long in the process of setting up to remember that (to read through the presentation where it says that).
> 
> Yeah, keep in mind that the main goal is just to get people comfortable with a Linux terminal. It's probably best not to do anything too fancy (or try to teach too much), otherwise, that might detract from the main goal.

I'll probably stick to what you had last year: "here is how to find downloaded files; here is how to put them on the server; this is what compiling someone else's code *should* look like; this is how to run it" is a pretty decent goal.

DWesl committed:
STY: Add pre-commit and run it on repository.

Removes trailing whitespace, ensures one newline at end of file,
adds encoding declaration to python files, and forces consistent
setup.cfg format.

This is just the auto-fixes; I still need to address the
warnings from flake8 and friends.

DWesl committed:
STY: Get pre-commit passing.

Mostly fixing flake8 warnings.

DWesl committed:
TST: Run example code in docstrings.

There's an example in interp.py to make the run pass.

DWesl committed:
FIX: Add __init__.py so python recognizes the package.

I thought this wasn't needed anymore.
